server/src/sas/formatter/parser.ts

```diff
@@ -0,0 +1,303 @@
+    return false;
+  }
+  while (block.outerBlock) {
+    if (block.outerBlock === parent) {
+      return true;
+    }
+    block = block.outerBlock;
+  }
+  return false;
+};
+
+const isStartingRegion = (parents: Region[], currentStatement: Statement) => {
+  if (/^%?do$/i.test(currentStatement.children[0].text)) {
+    return true;
+  }
+  if (
+    /^(%?if|%?else|when)$/i.test(currentStatement.children[0].text) &&
+    currentStatement.children.length > 1
+  ) {
+    for (let i = 1; i < currentStatement.children.length - 1; i++) {
+      if (/^%?do$/i.test(currentStatement.children[i].text)) {
+        return true;
+      }
+    }
+  }
+
+  let index = parents.length - 1;
+  while (index >= 0 && !parents[index].block) {
+    index--;
+  }
+  const parent = parents[index];
+  if (!parent || !parent.block) {
+    return false;
+  }
+
+  const map: Record<string, RegExp> = {
+    data: /^select$/i,
+    ds2: /^(select|method|data|package|thread)$/i,
+    template: /^(define|begingraph|layout)$/i,
+  };
+
+  const block = parent.block;
+  let procName = "";
+  if (block.name === "PROC") {
+    const firstToken = parent.children[0].children[1];
+    if (firstToken && "text" in firstToken) {
+      procName = firstToken.text;
+    }
+  }
+  const regexp = map[(procName || block.name).toLowerCase()];
+  if (regexp && regexp.test(currentStatement.children[0].text)) {
+    return true;
+  }
+  return false;
+};
+
+export const getParser =
+  (model: Model, tokens: Token[], syntaxProvider: SyntaxProvider) => () => {
+    const root: Program = {
+      type: "program",
+      children: [],
+    };
+    const parents: Region[] = [];
+    let region: Region | undefined = undefined;
+    let currentStatement: Statement | undefined = undefined;
+    let prevStatement: Statement | undefined = undefined;
+    for (let i = 0; i < tokens.length; i++) {
+      const node = tokens[i];
+      let parent = parents.length ? parents[parents.length - 1] : root;
+      if (node.type === "sec-keyword" || node.type === "macro-sec-keyword") {
+        const block = syntaxProvider.getFoldingBlock(
+          node.start.line,
+          node.start.column,
+          true,
+          true,
+          true,
+        );
+        if (
+          block &&
+          block.startLine === node.start.line &&
+          block.startCol === node.start.column
+        ) {
+          if (region && hasParent([...parents, region], block)) {
+            parents.push(region);
+            parent = region;
+          }
+          region = {
+            type: "region",
+            block: block,
+            children: [],
+          };
+          parent.children.push(region);
+        }
+      }
+      if (!currentStatement) {
+        currentStatement = {
+          type: "statement",
+          name: node.text,
+          children: [],
+        };
+        if (region) {
+          if (region.children.length === 0) {
+            // section start
+            if (prevStatement) {
```

Comment:
```suggestion
          if (region.children.length === 0 && prevStatement) {
```
Looks like this can _maybe_ be combined

Reply:
done in bd4862a